snowflake_stmt.c

```diff
@@ -13,56 +13,322 @@
 #include "zend_exceptions.h"
 
+/**
+ * Destroy a previously constructed statement object.
+ *
+ * This should do whatever is necessary to free up any driver specific 
+ * storage allocated for the statement. The return value from this function 
+ * is ignored.
+ * 
+ * @param stmt Pointer to the statement structure initialized by handle_preparer.
+ * @return 1 if success or 0 if error occurs
+ */
 static int pdo_snowflake_stmt_dtor(pdo_stmt_t *stmt) /* {{{ */
 {
+    PDO_DBG_ENTER("pdo_snowflake_stmt_dtor");
+    pdo_snowflake_stmt *S = stmt->driver_data;
+    pdo_snowflake_db_handle *H = S->H;
+
+    if (S->params) {
+        efree(S->params);
+    }
+    S->params = NULL;
+
+    if (S->data) {
+        efree(S->data);
+    }
+    S->data = NULL;
+
+    if (S->bound_result) {
+        efree(S->bound_result);
+    }
+    S->bound_result = NULL;
+
     PDO_DBG_RETURN(1);
 }
 /* }}} */
 
-static int pdo_snowflake_stmt_execute(pdo_stmt_t *stmt) /* {{{ */
+/**
+ * Execute a prepared statement. This is called by pdo_snowflake_stmt_execute.
+ * 
+ * Binds parameters and columns if required.
+ * 
+ * @param stmt Pointer to the statement structure initialized by handle_preparer.
+ * @return 1 if success or 0 if error occurs
+ */
+static int pdo_snowflake_stmt_execute_prepared(pdo_stmt_t *stmt) /* {{{ */
 {
+    PDO_DBG_ENTER("pdo_snowflake_stmt_execute_prepared");
+    int i;
+    pdo_snowflake_stmt *S = stmt->driver_data;
+    pdo_snowflake_db_handle *H = S->H;
+
+    // TODO: bind parameters
+    S->params = NULL;
+    // ecalloc((size_t)S->num_params, sizeof(SNOWFLAKE_BIND_INPUT));
+
+    /* execute */
+    if (snowflake_execute(S->stmt) != SF_STATUS_SUCCESS) {
+        pdo_snowflake_error_stmt(stmt);
+        PDO_DBG_RETURN(0);
+    }
+
+    /* Bind Columns/Results before fetching */
+    stmt->column_count = (int)snowflake_num_fields(S->stmt);
+    PDO_DBG_INF("number of columns: %d", stmt->column_count);
+    S->bound_result = ecalloc((size_t)stmt->column_count,
+                              sizeof(SNOWFLAKE_BIND_OUTPUT));
+    S->data = ecalloc((size_t)stmt->column_count, sizeof(void*));
+    for (i = 0; i < stmt->column_count; ++i)
+    {
+        size_t len = 0;
+        SNOWFLAKE_COLUMN_DESC* desc = S->stmt->desc[i];
+        S->bound_result[i].idx = (size_t)i+1;  /* 1 based index */
+        S->bound_result[i].type = SF_C_TYPE_STRING;
+        PDO_DBG_INF("prec: %d, scale: %d, nane: %s, type: %d",
+                    desc->precision, desc->scale, desc->name, desc->type);
+        if (desc->type == SF_TYPE_FIXED) {
+            if (desc->scale == 0) {
+                len = (size_t)desc->precision;
+            } else {
+                len = (size_t)desc->precision + desc->scale + 1;
+            }
+        } else if (desc->type == SF_TYPE_TEXT) {
+            len = 100;
+        }
+        S->bound_result[i].max_length = len + 1; // change this
+        S->data[i] = ecalloc(len+1, sizeof(char)); // TODO: change this
```

Comment:
Actually the null termination is not required for PHP. So I want to remove +1 here.